design_documents/parametrizing_tests_with_fixed_values.md

```diff
@@ -0,0 +1,143 @@
+# Parametrizing Tests With Fixed Values
+
+## Context
+
+Various test runners for other programming languages provide ways of parametrizing test cases with fixed values.
+This way, multiple test cases can be executed for the same test source code.
+
+## Goal
+
+Propose a way of adding support for parametrized tests to Starknet Foundry's `snforge`.
+
+## Proposed Solution
+
+Introduce new attribute `test_case`.
+
+Generate code for each of the test cases so no handling of values injection, etc. is necessary.
+The compiler would then validate if the values for the generated code are actually valid.
+Thanks to that, we could have failures for invalid arguments early in the execution.
+
+### Parametrization With Arguments
+
+```cairo
+#[test_case(1, 3)]
+#[test_case(3, 5)]
+fn my_test(a: felt252, b: u32) {
+    // ...
+}
+```
+
+For test cases like this, we would generate two actual test cases:
+
+```cairo
+#[test]
+fn my_test_case_1() {
+  let a: felt252 = 1;
+  let b: u32 = 3;
+  // ...
+}
+
+#[test]
+fn my_test_case_2() {
+    let a: felt252 = 3;
+    let b: u32 = 5;
+  // ...
+}
+```
+
+### Generated Cases Names
+
+Generated test cases should be named in such a way that we can automatically detect which tests are generated cases of
+the same base test.
+
+Test in `snforge` can be filtered by name.
+It is important that generated test names do not break the test filtering logic.
+
+This can be resolved by either using names for generated tests that still work with filters as the base test would do,
+or changing the filtering logic, so it can recognize generated test cases and treat them accordingly.
+
+### Parametrizing With Complex Types
+
+Complex types would result in the same code generation as simple types:
+
+```cairo
+#[test_case(MyStruct { a: 1, b: array![2, 3] })]
+fn my_test(a: MyStruct) {
+    // ...
+}
+```
+
+```cairo
+#[test]
+fn my_test_1() {
+    let a = MyStruct { a: 1, b: array![2, 3] };
+    // ...
+}
+```
+
+### Problems With Code Generation
+
+The code generation solution has some problems that will need to be addressed before it is implemented.
+See [section below](#possible-problems-with-code-generation) for details.
+
+## User Experience
+
+Unlike fuzz-tests, we should run and indicate the result of all parametrized cases.
+If some test cases fail, others should still be executed and their results should be displayed to the user.
+
+An example output could look similarly to this:
+
+```shell
+$ snforge test
+[PASS] tests::parametrized(a = 1, b = 2, c = my_fixture(1, 2))
```

Comment:
just noting that there is no information about fixtures up to this moment. Idk if it is ok or not

Reply:
Leftover from previous version, removed.